orbtrace/debug/cmsis_dap.py

```diff
@@ -487,7 +487,7 @@ def RESP_JTAG_Configure_Setup(self, m):
         m.d.sync += [
             self.dbgif.command.eq( CMD_SET_JTAG_CFG ),
             self.jtag_ircount.eq(0),
-            self.dbgif.dev.eq(self.rxBlock.bit_select(8,3)-1)
+            self.dbgif.dev.eq(self.rxBlock.bit_select(8,4)-1)
```

Comment:
dev is a 3 bit field in the jtagIF at the moment?  That was an arbitary choice, we can make it 4 if its useful, but it needs to be changed in a few places if so (+other things that use it)

Reply:
I don't see anywhere in the CMSIS-DAP spec that specifies how many devices in a JTAG chain are supported apart from the absolute upper limit of 256 due to the index being an 8-bit field.

FWIW the reference implementation also defaults to max 8 devices: https://github.com/ARM-software/CMSIS/blob/master/CMSIS/DAP/Firmware/Config/DAP_config.h#L65

Reply:
the thinking here is that dev can stay a 3-bit field but we must select *4* bits for the subtraction so that `16 - 1` gives the correct answer without problem. We can then discard the top bit after the subtraction.

It should be noted we should really also detect and handle the case the number of devices is sent as 0 and handle that as right now it probably makes ORBTrace think there are 16 devices on the chain.

Reply:
Ah, now I follow better. Still, as long as the result is going into a 3-bit field it doesn't really matter if you're doing a 3-bit or a 4-bit subtraction, because the difference only affects the bit that's getting discarded anyway. If we're getting the maximum valid count of 8, pick the bottom three bits and subtract one, we're effectively doing 0 - 1 which underflows and becomes 7.

For range checking, we should consider the entire 8-bit field and check that the count is both larger than zero and no larger than 8.

Reply:
That should be ok. This value is used to set the number of devices in the chain-1, so 1 -> 0, 8 -> 7 etc. For the max'es we've got now it doesn't make any difference though.